lib/Parser/Parse.cpp

```diff
@@ -10037,14 +10037,35 @@ ParseNodePtr Parser::ParseStatement()
         }
 
         Assert(pnode != nullptr);
-        ParseNodeFnc* pNodeFnc = (ParseNodeFnc*)pnode;
+
         if (labelledStatement)
         {
             if (IsStrictMode())
             {
                 Error(ERRFunctionAfterLabelInStrict);
             }
-            else if (pNodeFnc->IsAsync())
+            // #sec-with-statement-static-semantics-early-errors states that the Statement of
+            // a WithStatement throws a Syntax Error if the Statement is a LabelledFunction.
+            else if (m_pstmtCur && m_pstmtCur->pnodeStmt && m_pstmtCur->pnodeStmt->nop == knopWith)
+            {
+                Error(ERRStmtOfWithIsLabelledFunc);
+            }
+
+            ParseNodeFnc* pNodeFnc;
+
+            // pnode can be a knopBlock due to ParseFncDeclCheckScope, which
+            // can return a ParseNodeBlock that contains a ParseNodeFnc.
+            if (pnode->nop == knopBlock && ((ParseNodeBlock*)pnode)->pnodeStmt
```

Comment:
Can we use `pnode->AsParseNodeBlock()` instead of casting to `ParseNodeBlock*`? Maybe like this or something?

```c++
if (pnode->nop == knopBlock)
{
  ParseNodeBlock* pNodeBlock = pnode->AsParseNodeBlock();
  if (pNodeBlock->pnodeStmt && pNodeBlock->pnodeStmt->nop == knopFncDecl)
  {
    pNodeFnc = pNodeBlock->pnodeStmt->AsParseNodeFnc();
  }
}
```